sklearn/decomposition/dict_learning.py

```diff
@@ -545,6 +545,230 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+        Whether to also return the code U or just the dictionary V.
+
+    dict_init : array of shape (n_components, n_features),
+        Initial value for the dictionary for warm restart scenarios.
+
+    callback :
+        Callable that gets invoked every five iterations.
+
+    batch_size : int,
+        The number of samples to take in each batch.
+
+    verbose :
+        Degree of output the procedure will print.
+
+    shuffle : boolean,
+        Whether to shuffle the data before splitting it in batches.
+
+    n_jobs : int,
+        Number of parallel jobs to run, or -1 to autodetect.
+
+    method : {'lars', 'cd'}
+        lars: uses the least angle regression method to solve the lasso problem
+        (linear_model.lars_path)
+        cd: uses the coordinate descent method to compute the
+        Lasso solution (linear_model.Lasso). Lars will be faster if
+        the estimated components are sparse.
+
+    iter_offset : int, default 0
+        Number of previous iterations completed on the dictionary used for
+        initialization.
+
+    random_state : int or RandomState
+        Pseudo number generator state used for random sampling.
+
+    return_n_iter : bool
+        Whether or not to return the number of iterations.
+        Overridden by ``return_code``.
+
+    Returns
+    -------
+    code : array of shape (n_samples, n_components),
+        the sparse code (only returned if `return_code=True`)
+
+    dictionary : array of shape (n_components, n_features),
+        the solutions to the dictionary learning problem
+
+    n_iter : int
+        Number of iterations run. Returned only if `return_n_iter` is
+        set to `True`.
+
+    See also
+    --------
+    dict_learning
+    SparseHebbianLearning
+    DictionaryLearning
+    MiniBatchDictionaryLearning
+    SparsePCA
+    MiniBatchSparsePCA
+
+    """
+    if n_components is None:
+        n_components = X.shape[1]
+
+    t0 = time.time()
+    n_samples, n_features = X.shape
+    random_state = check_random_state(random_state)
+
+    if n_jobs == -1:
+        n_jobs = cpu_count()
+
+    dictionary = random_state.randn(n_features, n_components)
+    norm = np.sqrt(np.sum(dictionary**2, axis=1)).T
+    dictionary /= norm[:, np.newaxis]
+    norm = np.sqrt(np.sum(dictionary**2, axis=1)).T
+
+    if verbose == 1:
+        print('[dict_learning]', end=' ')
+    gain = np.ones(n_components)
+    gain_ = np.ones(n_components)
+
+    n_batches = ceil(float(len(X)) / batch_size)
+    if shuffle:
+        X_train = X.copy()
+        random_state.shuffle(X_train)
+    else:
+        X_train = X
+    batches = np.array_split(X_train, n_batches)
+    batches = itertools.cycle(batches) # Return elements from list of batches until it is exhausted. Then repeat the sequence indefinitely.
+
+    # If n_iter is zero, we need to return zero.
+    ii = iter_offset - 1
+
+    for ii, this_X in zip(range(iter_offset, iter_offset + n_iter), batches):
+        dt = (time.time() - t0)
```

Comment:
this bit of code is not related to this PR and is repeated for instance in ``dict_learning``, ``dict_learning_online `` and ``dict_learning_grad``